daml/ContingentClaims/Observation.daml

```diff
@@ -7,140 +7,111 @@
 
 module ContingentClaims.Observation (
     Observation(..)
-  , observe
-  , expiry
-  , eval
-  , HasFixings(..)
-  , Key
 ) where
 
 import ContingentClaims.Observable qualified as O
-import Daml.Control.Arrow (Kleisli(..))
-import Daml.Control.Category
+import ContingentClaims.Recursion (cataM)
+import Daml.Control.Arrow (Kleisli(..), arr)
+import Daml.Control.Recursion (Recursive(..), Corecursive(..), project, embed)
+import DA.Traversable (Traversable(..))
+import DA.Foldable (Foldable(..))
+import Prelude hiding (key, (.))
 
-type Key = Text
 type T = Observation
 
 -- | Concrete implementation of `Observable`, which can be serialized.
-data Observation t a
-  = DecimalConst Decimal
-  | DecimalObs Key
-  | DecimalLte (Observation t Decimal, Observation t Decimal)
-  | DecimalEqu (Observation t Decimal, Observation t Decimal)
-  | DecimalAdd (Observation t Decimal, Observation t Decimal)
-  | DecimalNeg (Observation t Decimal)
-  | DecimalMul (Observation t Decimal, Observation t Decimal)
-  | DecimalDiv (Observation t Decimal, Observation t Decimal)
-  | BoolConst Bool
-  | DateIdentity
-  | DateConst t
-  | DateLte (Observation t t, Observation t t)
-  | DateEqu (Observation t t, Observation t t)
+-- Conceptually it's helpful to think of this as the type `t -> x`, or `t -> Update x`.
+data Observation t x
+  = Const with value: x 
+    -- ^ A named constant
+  | Observe with key: O.Key 
+    -- ^ A named parameter
```

Comment:
```suggestion
data Observation t a x
  = Param (Either a x )
    -- ^ A named parameter
```

This would have two advantages: 
1. We'd be able to write proper bi-functor instances for Observable.
2. We could tie the key to the asset at the `Claim` level. The reason I've avoided that and used a `type Key = Text` so far, is that as @georg-da pointed out, we might want to distinguish between 'close', 'high' 'low', etc. (but this could be worked around by wrapping the asset in an effect, that specifies what type of data to fetch). Also, not all parameters are necessarily 'assets' e.g. temperature etc.